### PR TITLE
Add export-all command

### PR DESCRIPTION
**Pull request by sztomi**
Changelog: Feature: This PR adds an `export-all` command that allows exporting a monorepo of recipes quickly. For comparison, we are currently using an external script that launches multiple `conan export` commands in parallel. Because a no-op conan call is already about 500ms (command.py is HUGE), we repeatedly spend a lot of time redoing the same things in conan startup. The idea is to only do these things once and then repeatedly (in parallel) export the recipes.

This yields a 19x speedup for the same monorepo (containing 83 recipes):

```
❯ hyperfine -w 3 'rm -rf ./conan_home/.conan/data && ./dep_builder.py -o'
Benchmark #1: rm -rf ./conan_home/.conan/data && ./dep_builder.py -o
  Time (mean ± σ):     34.395 s ±  0.660 s    [User: 190.776 s, System: 11.921 s]
  Range (min … max):   33.169 s … 35.532 s    10 runs

❯ hyperfine -w 3 'rm -rf ./conan_home/.conan/data && python -m conans.conan export-all packages plex/stable'
Benchmark #1: rm -rf ./conan_home/.conan/data && python -m conans.conan export-all packages plex/stable
  Time (mean ± σ):      1.879 s ±  0.039 s    [User: 1.773 s, System: 0.957 s]
  Range (min … max):    1.826 s …  1.945 s    10 runs
```

This is a draft PR because this implementation is the bare minimum:

  * the output from each parallel run is interleaved
  * the threadpool is hardcoded to 8 threads

I'd like to get your opinion on this before I continue and solve the above. Would you be willing to add a command like this?

Docs: https://github.com/conan-io/docs/pull/XXXX

- [ ] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [ ] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [ ] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/client/command.py [hunk 1]
```diff
@@ -1066,6 +1066,49 @@ def export(self, *args):
...
+        """
+        parser = argparse.ArgumentParser(description=self.export.__doc__,
+                                         prog="conan export",
+                                         formatter_class=SmartFormatter)
+        parser.add_argument("path", help=_PATH_HELP)
+        parser.add_argument("reference", nargs='?', default=None,
+                            help="user/channel, Pkg/version@user/channel (if name "
+                                 "and version are not declared in the conanfile.py) "
+                                 "Pkg/version@ if user/channel is not relevant.")
+        parser.add_argument('-k', '-ks', '--keep-source', default=False, action='store_true',
```
Comment (memsharded) [inline conans/client/command.py:1083]: Is the --keep-source very necessary for this use case? Maybe not? I would try to simplify and remove it.

File: conans/client/command.py [hunk 2]
```diff
@@ -1066,6 +1066,49 @@ def export(self, *args):
...
+        """
+        Exports multiple recipes from a monorepo.
+
+        Each subdirectory of the given path will be considered a reciped folder.
+        """
+        parser = argparse.ArgumentParser(description=self.export.__doc__,
+                                         prog="conan export",
+                                         formatter_class=SmartFormatter)
+        parser.add_argument("path", help=_PATH_HELP)
+        parser.add_argument("reference", nargs='?', default=None,
```
Comment (memsharded) [inline conans/client/command.py:1079]: Reference is unused

File: conans/client/conan_api.py [hunk 3]
```diff
@@ -842,6 +842,37 @@ def export(self, path, name, version, user, channel, keep_source=False, cwd=None
...
+            graph_lock = graph_lock_file.graph_lock
+            self.app.out.info("Using lockfile: '{}'".format(lockfile))
+
+        self.app.load_remotes()
+
+        from concurrent.futures import as_completed, ThreadPoolExecutor
+        from pathlib import Path
+
+        futures = []
+        path = Path(path)
```
Comment (memsharded) [inline conans/client/conan_api.py:861]: We probably want to keep the code Py2 compatible, it won't cost much to use old style ``os.path`` syntax to work.

File: conans/client/conan_api.py [hunk 4]
```diff
@@ -842,6 +842,37 @@ def export(self, path, name, version, user, channel, keep_source=False, cwd=None
...
+        conanfiles = [_get_conanfile_path(str(subdir.absolute()), cwd, py=True) for subdir in path.iterdir()]
+        with ThreadPoolExecutor(max_workers=8) as pool:
```
Comment (memsharded) [inline conans/client/conan_api.py:863]: Use configurable ``cpu_count()`` as parameter.

**Review by memsharded — COMMENTED** [review]
The feature would need thorough, comprehensive testing:

- Making sure resulting lockfiles are consistent.
- ignore-dirty cases

**Comment by sztomi** [comment]
After some discussion in slack DMs, I'll be closing this. We will instead use the (experimental) conan API to implement the same.

**Pull request closed without merging**